### ISSUE TITLE
Terminal hardware is always left in application mode (keypad_xmit) after running a console program

### ISSUE THREAD

**Issue by kkm000** [dotnet/corefx#32805]
This problem has been annoying me for quite a while, and I worked around it by emitting the `rmkx` terminfo sequence in my bash prompt. But I still think it makes sense to report it. Basically, after (almost?) any `dotnet` command that involves CLR console I/O, the terminal emulator is left in the application mode (`keypad_xmit`), messing my command line handling (e. g. arrows move by word `<ESC> O C`, not by character `<ESC> C`, in readline bash prompt, etc. -- probably because of my customized `.inputrc`, designed to work with both VT100 and rxvt style emulators). Other programs behave nice; e. g., just typing `man man`, or invoking `less` or `vim` and then exiting them reverts keyboard to local keypad mode. Just for reference, this is a hopelessly headless Ubuntu 18.04 machine:
```
$ uname -a
Linux yupana 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:39:52 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ echo $TERM
xterm-256color
$ dotnet --version
2.1.403
```
I noticed by using a script(1) dump that the `smkx` aka `keypad_xmit` code is emitted more than once, but then there is no reverting `rmkx` aka `keypad_local` is emitted ever. Here is, for example, how captured prologue of `dotnet build` looks (not marking literal linefeeds), that emits it twice:
```
Script started on 2018-10-12 23:02:55-0700
. . .
dotnet build -c Release -v:n
<ESC>[?1h<ESC>=Microsoft (R) Build Engine version 15.8.169+g1ccb72aefa for .NET Core

Copyright (C) Microsoft Corporation. All rights reserved.

<ESC>[?1h<ESC>=Build started 10/12/18 11:00:40 PM.
```
I just attempted to trace the issue down by eyeballing the console handling code. My terminal emulator is pretending to be xterm-256color compatible (MobaXTerm), and infocmp(1) does show entries for both sequences for this term type:
```
rmkx=\E[?1l\E>
smkx=\E[?1h\E=
```
The terminfo entries have manifest numeric IDs in `/usr/include/term.h
```
#define keypad_local                   CUR Strings[88]
#define keypad_xmit                    CUR Strings[89]
```
but the index 88 is not even in the [`enum WellKnownStrings` in TermInfo.cs](https://github.com/dotnet/corefx/blob/14072950a/src/System.Console/src/System/TermInfo.cs#L61) file that reads the terminfo file directly (but `KeypadXmit = 89` is). Also, it seems that [pal_console.c](https://github.com/dotnet/corefx/blob/14072950a/src/Native/Unix/System.Native/pal_console.c) does a good job of restoring the application mode at a few points by sending the `keypad_xmit` sequence, apparently since the MR #6488 fixing the issue #5902, but unless I misunderstand it, never attempts to send a matching `keypad_local` in any of its Uninitialize() functions, where tty driver attributes are restored.

**Comment by kkm000** [comment]
@danmosemsft, a fair question! I do not know, really. If I had the whole thing checked out and churning and able to test, I'd implement and send a fix in an hour. The missing piece looks simple to integrate, would touch only `internal` type signatures and use already present signal-safe routines in the native shim (I hope I use the correct term).

The big deal is I know practically nothing how the whole stack works. I just found this repo by searching for a type named "ConsolePal" that popped up in a stack backtrace that a c# program using Console.ReadKey() so helpfully puked when I invoked it with "</dev/null". It gave me a hint where to look for a low-level library responsible for tty I/O, that's it. The rest just some common sense and a bit of reading the code.

I mean, I could certainly try to check out and build the repo; I routinely develop in C++ on both Windows and Linux and C# on Windows. But I have no idea what would I do with the libraries that I build. I know little what is under the hood of the "dotnet" command. I quickly read the [documentation on building an testing](https://github.com/dotnet/corefx/tree/master/Documentation/building), but I do not understand how I could incorporate what I build into the rest of the stack. Do I also need to build the coreclr repo, or can I run an application with my own build of corefx but the existing RTM coreclr runtime installed from a .deb? If you could point me to any docs or explain that in a few words, I would certainly give it a try.

**Comment by stephentoub** [comment]
@kkm000, thanks. Is there a way to detect whether the current setting is already keypad_xmit?  It should be fairly straightforward to fix up the uninitialize routine to output keypad_local, but that could itself cause a potential issue if the terminal were already set as keypad_xmit when the process was launched, as we'd then be switching away from what the environment described.

**Comment by kkm000** [comment]
@stephentoub:

>  Is there a way to detect whether the current setting is already keypad_xmit?

I do not believe there is, and most probably not without running into another timeout problem like #31517. But it's not really necessary. For all I know, no sane program would assume that the terminal is in the keypad mode when it starts. And I know for sure that bash (rather readline) assumes it is not :(

My understanding is that the "local" mode is the default, and it's always ok to revert to it. [vim does this unconditionally](https://github.com/vim/vim/blob/64f410742/src/term.c#L3578) upon exit (here [`T_KE` corresponds](https://github.com/vim/vim/blob/64f410742/src/term.c#L1603) to yet another name for `keypad_local`, the termcap `ks`/`ke` pair). [less does it pretty much always](https://github.com/vbwagner/less/blob/7b48019d/screen.c#L1575) as well (when did you last type `less --no-init`? :) ).

[vim documentation has a little bit](https://github.com/vim/vim/blob/64f410742/runtime/doc/term.txt#L81) on its keypad mode control.
